### ISSUE TITLE
Add support for iam_metadata within aws identity integration

### ISSUE THREAD

**Issue by giogranda85** [hvac/hvac#1029]
There is currently a gap between what is documented in [HVAC](https://hvac.readthedocs.io/en/stable/source/hvac_api_auth_methods.html#hvac.api.auth_methods.Aws.configure_identity_integration) vs [Vault's API](https://developer.hashicorp.com/vault/api-docs/auth/aws#configure-identity-integration). There is no parameter available to adjust the iam_metadata or ec2_metadata within configure_identity_integration.

**Comment by danholodak** [comment]
Hi! I'd like to take a stab at this, but I'm not sure what kind of validation makes sense to add for the 2 new parameters (`iam_metadata` + `ec2_metadata`).

The existing 2 params check the value against a list ([code](https://github.com/hvac/hvac/blob/6e47552b12573bb150a042922031bf95e4231759/hvac/api/auth_methods/aws.py#L159-L172)), but I can't tell if the endpoint we're calling will actually fail if there are invalid entries in the arrays for the 2 new ones or if it just ignores them, and I can't figure out how to test it. When I try to hit the endpoint locally (http://127.0.0.1:8200/v1/auth/aws/config/identity) I get a `route entry not found` error.

I would appreciate any advice on either:
-  how to test what the response to invalid input is or
- if you would advise defaulting to validating (making sure everything in a passed array is included in vault's list) or not (just confirming that either an array or the string "default" is passed)

Thanks!

**Comment by briantist** [comment]
@danholodak welcome, thanks for your interest!

Normally I would recommend adding tests for this to the integration tests, but... we don't have them! Some integrations like this one are difficult to test, because they interact with an external service (AWS) which we don't have access to in our CI.

This also makes it difficult to test locally. If you have access to an AWS account, and want to configure a local Vault against to it to test, that would be amazing, but not required.

The `route entry not found` error is because you first have to enable AWS auth in your local Vault before you can configure it.

You _might_ be able to get away with simply "enabling" the method and then you could call the identity config endpoint to a degree where you can get feedback on the parameter validity, that would be nice. If so, it would even be enough to add some tests!

The CLI command to enable AWS auth:
```
vault auth enable aws
```
see also: https://developer.hashicorp.com/vault/docs/auth/aws#via-the-cli

---

> if you would advise defaulting to validating (making sure everything in a passed array is included in vault's list) or not (just confirming that either an array or the string "default" is passed)

As for validating the actual entries: I don't think we should do that in this library. We don't know if/when the list is going to change. If Vault adds a new valid entry for example, and then we don't realize it, our library gets in the way of anyone using the new value (and all previously released `hvac` versions cease to be usable for that use case for newer Vault versions).

So instead, I prefer we pass it along to the API, and then properly handle/report any resulting error.
The work you're doing to try to understand how Vault will respond to invalid input in those parameters is perfect.

Let me know if you get stuck and I'll see how I can unblock you, thanks again!

**Comment by danholodak** [comment]
Enabling AWS auth on command line worked perfectly - I didn't need to connect an AWS account to test if different params/values error'd or succeeded. Thanks!

It makes sense that this library shouldn't be responsible for validating the input - would it make sense to remove the existing validation for the `*_alias` parameters (e.g. `ALLOWED_IAM_ALIAS_TYPES`) in that case?

For adding a test - would that be creating a `test_aws.py` file in the `integration_tests/v1` folder? How would I do the equivalent of running `vault auth enable aws` in a test?

**Comment by briantist** [comment]
> Enabling AWS auth on command line worked perfectly - I didn't need to connect an AWS account to test if different params/values error'd or succeeded. Thanks!

🎉 great!

> It makes sense that this library shouldn't be responsible for validating the input - would it make sense to remove the existing validation for the `*_alias` parameters (e.g. `ALLOWED_IAM_ALIAS_TYPES`) in that case?

I would say let's leave it alone for now. That could be done in a separate PR maybe, especially if there were tests, but not necessary at the moment.

> For adding a test - would that be creating a `test_aws.py` file in the `integration_tests/v1` folder? How would I do the equivalent of running `vault auth enable aws` in a test?

We have both integration and unit tests. For integration in this case, it'd be in here with the other auth methods: https://github.com/hvac/hvac/tree/main/tests/integration_tests/api/auth_methods

For unit tests, https://github.com/hvac/hvac/tree/main/tests/unit_tests/api/auth_methods

For unit tests you can see that we usually mock the responses to the API call (these tests don't actually run against a Vault server, so no actual enable call is ever made):
- https://github.com/hvac/hvac/blob/main/tests/unit_tests/api/auth_methods/test_azure.py
- https://github.com/hvac/hvac/blob/main/tests/unit_tests/api/auth_methods/test_gcp.py

For integration tests, you write `setUp` and `tearDown` methods that handle enabling and usually configuring the auth method before the tests start, and then undoing all that when the tests are finished:
- https://github.com/hvac/hvac/blob/main/tests/integration_tests/api/auth_methods/test_azure.py
- https://github.com/hvac/hvac/blob/main/tests/integration_tests/api/auth_methods/test_github.py

---

Before starting to write tests, I recommend getting your local environment set up so that you can run the existing tests locally, that way you'll have a tight feedback loop on your efforts. Pushing changes to run it in CI will be very slow, because as a new contributor, I have to manually approve CI runs for all of your pushes.

There's some info on setting up the local dev environment in the [Contributing doc](https://hvac.readthedocs.io/en/stable/contributing.html) but some of that doc may need updating, so if you run into any issues, it would be a good opportunity for us to update that too.

Let me know if you need any help getting your local environment working.

**Comment by danholodak** [comment]
I opened a PR for this here: https://github.com/hvac/hvac/pull/1125. Please let me know if it needs anything else, when you get a chance :)
